### ISSUE TITLE
std::chrono::high_resolution_clock seems to be inaccurate in Linux

### ISSUE THREAD

**Issue by jamshark70** [supercollider/supercollider#4579]
## Environment

* SuperCollider version: 3.10.3
* Operating system: Ubuntu Studio 18.04

## Steps to reproduce

https://www.listarc.bham.ac.uk/lists/sc-users/msg65580.html

Evidence from this thread strongly suggests that std::chrono::high_resolution_clock does not exhibit accurate timing in Linux, and that it is significantly less inaccurate in Mac and Windows.

The inaccuracy appears to be proportional to the number of times per second that the clock is asked to awake. If I run a 12 second test with 10 events per second, sclang loses about 0.145 seconds (about 10 ms deviation per second, or 1% slower). If I run the same test with 100 events per second, sclang loses about 0.480 seconds (about 40 ms/sec, or 4% slower).

So the effect is minimal for typical musical contexts, where 20-30 events per second would cover most cases. But it does mean that high-density granular synthesis with one grain per synth is not reliable in Linux presently.

Also, the effect is cumulative: the total drag is worse for a 12 second test compared to a 2 second test.

If the effect were networking jitter, I would expect some of the measured times to be shorter than expected and others to be longer, with the average being close to correct. But all of the measurements skew long. The only explanation that makes any sense is that TempoClock, in Linux, runs slower than real time. TempoClock relies on std::chrono::high_resolution_clock, leading to the conclusion that std::chrono does not work well in Linux.

```
// ROUTINE TEST
(
// uses interpreter variables, not safe to run concurrently
s.waitForBoot {
	var recSynth;
	var factor = 100;
	b = Buffer.alloc(s, s.sampleRate * 12, 1);
	SynthDef("help-OffsetOut", { arg out = 0, freq = 440, dur = 0.005;
		var env;
		env = EnvGen.ar(Env.perc(0.0001, dur, 0.2), doneAction:2);
		OffsetOut.ar(out, SinOsc.ar(freq, 0, env))
	}).send(s);
	SynthDef(\rec, { |out, bufnum, bus|
		var rec, done;
		rec = RecordBuf.ar(In.ar(bus, 1), bufnum, loop: 0, doneAction: 2);
		Line.kr(0, 1, b.duration, doneAction: 2);  // FFS
	}).send(s);
	s.sync;

	s.makeBundle(0.2, {
		recSynth = Synth(\rec, [bufnum: b, bus: 0]);
	});

	(b.duration * factor * 0.95).do {
		s.sendBundle(0.2, ["/s_new", "help-OffsetOut", -1, 2, recSynth.nodeID]);
		factor.reciprocal.wait;
	};

	// exited loop, recording should be done
	b.getToFloatArray(0, b.numFrames, wait: -1, action: { |data| d = data });
	// 'forkIfNeeded' in that method, should be all finished
	// scan for onsets
	o = List.new;
	(2 .. d.size-1).do { |i|
		if(d[i-2] == 0 and: { d[i-1] == 0 and: { d[i] != 0 } }) {
			o.add(i);
		}
	};

	p = o.differentiate.drop(1);
	[p.minItem, p.maxItem, p.mean, p.median].postln;

	b.free;
};
)

// IMPULSE TEST
(
// uses interpreter variables, not safe to run concurrently
s.waitForBoot {
	var recSynth;
	var factor = 100;
	b = Buffer.alloc(s, s.sampleRate * 12, 1);
	SynthDef("help-OffsetOut", { arg out = 0, freq = 440, dur = 0.005;
		var env;
		var trig = Impulse.ar(factor),
		count = PulseCount.ar(trig),
		num = b.duration * factor * 0.95;
		FreeSelf.kr(count >= num);
		env = EnvGen.ar(Env.perc(0.0001, dur, 0.2), trig);
		OffsetOut.ar(out, SinOsc.ar(freq, 0, env))
	}).send(s);
	SynthDef(\rec, { |out, bufnum, bus|
		var rec, done;
		rec = RecordBuf.ar(In.ar(bus, 1), bufnum, loop: 0, doneAction: 2);
		Line.kr(0, 1, b.duration, doneAction: 2);  // FFS
	}).send(s);
	s.sync;

	s.makeBundle(0.2, {
		recSynth = Synth(\rec, [bufnum: b, bus: 0]);
	});

	s.sendBundle(0.2, ["/s_new", "help-OffsetOut", -1, 2, recSynth.nodeID]);
	b.duration.wait;

	// exited loop, recording should be done
	b.getToFloatArray(0, b.numFrames, wait: -1, action: { |data| d = data });
	// 'forkIfNeeded' in that method, should be all finished
	// scan for onsets
	o = List.new;
	(2 .. d.size-1).do { |i|
		if(d[i-2] == 0 and: { d[i-1] == 0 and: { d[i] != 0 } }) {
			o.add(i);
		}
	};

	p = o.differentiate.drop(1);
	[p.minItem, p.maxItem, p.mean, p.median].postln;

	b.free;
};
)


p.size
o.size


// factor = 10
// impulse
o.maxItem - o.minItem  // 498330, 114
// routine
o.maxItem - o.minItem  // 504751, 114

(504751 - 498330) / 44100 = 0.1456

// factor = 100
// impulse
o.maxItem - o.minItem  // 502299, 114
// routine
o.maxItem - o.minItem  // 523470, 114

(523470 - 502299) / 44100 = 0.4801
```

## Expected vs. actual behavior

While we don't promise sample accuracy in real time, we also observe that the deviation in inter-offset intervals (measured in samples) is:

- Mac/Windows: +/- 2 or 3 samples
- Linux: +2 to +60 samples (measurements on some other machines might go negative but still exhibit much wider deviation than the other platforms)

This is a rather poor result to report to Linux audio communities. I don't think we necessarily have to match Mac/Windows accuracy but this is a bit too much to pass without comment.

**Comment by brianlheim** [comment]
So, I'm a little confused about what the report is here. It sounds like the assertion is that in this code,

```supercollider
	(b.duration * factor * 0.95).do {
		s.sendBundle(0.2, ["/s_new", "help-OffsetOut", -1, 2, recSynth.nodeID]);
		factor.reciprocal.wait;
	};
```

the time interval between bundles being processed by the server is greater than `factor.reciprocal`. Isn't that expected behavior? If I'm misunderstanding, perhaps you could clean up the ticket description a little bit. There is a lot of code here and the email thread was hard to follow.

> TempoClock relies on std::chrono::high_resolution_clock, leading to the conclusion that std::chrono does not work well in Linux.

What does 'not work well' mean? If this is a problem with `std::chrono::high_resultion_clock`, it should be fairly straightforward to come up with a minimum reproducer (i.e., a c++ code) to demonstrate it; I would consider that solid evidence. 

Btw, it would be more accurate to say "libstdc++'s std::chrono::high_resolution_clock"; there are several implementations of the C++ standard library on Linux. I am assuming you mean the one that ships with gcc and that most people will compile with, but it always helps to clarify that.

**Comment by jamshark70** [comment]
Hm, you're right that there are more variables at play than I let on, and I might be jumping to a conclusion.

I think I'd better close this one and rewrite.

**Comment by sonoro1234** [comment]
Also supernova sampleclock should be used in test to discard some things!!!
